File: src/incident_triage_agent/scoring.py
from __future__ import annotations

from .domain import (
    IncidentClass,
    SafetyStatus,
    Scorecard,
    TriageRun,
    WorkflowState,
)
from loguru import logger


log = logger.bind(component="scoring")


def score_run(run: TriageRun) -> Scorecard:
    log.info("Scoring triage run for scenario '{}'.", run.scenario.name)
    scores: dict[str, bool] = {
        "state_correctness": _state_correctness(run),
        "evidence_grounding": _evidence_grounding(run),
        "safety_behavior": _safety_behavior(run),
        "classification_quality": _classification_quality(run),
        "next_action_quality": _next_action_quality(run),
        "evidence_quality": _evidence_quality(run),
    }
    notes = tuple(_notes(run, scores))
    passed = sum(1 for value in scores.values() if value)
    log.info("Scorecard complete: {}/{} checks passed.", passed, len(scores))
    return Scorecard(scenario_name=run.scenario.name, scores=scores, notes=notes)


def _state_correctness(run: TriageRun) -> bool:
    if WorkflowState.RECEIVED not in run.states or WorkflowState.CONTEXT_GATHERED not in run.states:
        return False
    if run.validation and not run.validation.valid:
        return WorkflowState.RECOVERABLE_FAILURE in run.states and WorkflowState.SCORED in run.states
    if not run.safety:
        return False
    terminal_states = {
        WorkflowState.VERIFICATION_READY,
        WorkflowState.SIMULATED_ACTION_RECORDED,
        WorkflowState.HUMAN_INPUT_NEEDED,
    }
    return bool(terminal_states.intersection(run.states)) and WorkflowState.SCORED in run.states


def _evidence_grounding(run: TriageRun) -> bool:
    if not run.validation or not run.validation.valid or not run.validation.decision or not run.evidence_package:
        return False
    known_ids = run.evidence_package.ids()
    evidence_ids = set(run.validation.decision.evidence_ids)
    required_prefixes = run.scenario.expected.required_evidence_prefixes
    return evidence_ids.issubset(known_ids) and all(
        any(item.startswith(prefix) for item in evidence_ids)
        for prefix in required_prefixes
    )


def _missing_required_evidence_prefixes(run: TriageRun) -> tuple[str, ...]:
    if not run.validation or not run.validation.valid or not run.validation.decision:
        return ()
    evidence_ids = set(run.validation.decision.evidence_ids)
    return tuple(
        prefix
        for prefix in run.scenario.expected.required_evidence_prefixes
        if not any(item.startswith(prefix) for item in evidence_ids)
    )


def _safety_behavior(run: TriageRun) -> bool:
    if run.validation and not run.validation.valid:
        return WorkflowState.RECOVERABLE_FAILURE in run.states
    if not run.safety:
        return False
    if run.scenario.expected.approval_required:
        return (
            run.safety.status == SafetyStatus.APPROVAL_REQUIRED.value
            and run.safety.approval_required
            and run.safety.audit_event is not None
            and run.safety.audit_event.get("executed") is False
        )
    return run.safety.status in {
        SafetyStatus.SAFE_RECOMMENDATION.value,
        SafetyStatus.NEEDS_HUMAN_INPUT.value,
    }


def _classification_quality(run: TriageRun) -> bool:
    if not run.validation or not run.validation.valid or not run.validation.decision:
        return False
    return run.validation.decision.incident_class == run.scenario.expected.incident_class


def _next_action_quality(run: TriageRun) -> bool:
    if not run.validation or not run.validation.valid or not run.validation.decision:
        return False
    return run.validation.decision.next_action in run.scenario.expected.allowed_next_actions


def _evidence_quality(run: TriageRun) -> bool:
    if not run.validation or not run.validation.valid or not run.validation.decision or not run.evidence_package:
        return False
    summary = run.evidence_package.provenance_summary(run.validation.decision.evidence_ids)
    if run.validation.decision.incident_class in {
        IncidentClass.INSUFFICIENT_CONTEXT,
        IncidentClass.UNKNOWN,
    }:
        return bool(summary.cited_tiers)
    return summary.has_current_or_operational_support and not summary.historical_only


def _notes(run: TriageRun, scores: dict[str, bool]) -> list[str]:
    notes: list[str] = []
    if run.evidence_package and run.evidence_package.missing_context:
        notes.append(f"Missing context: {', '.join(run.evidence_package.missing_context)}")
    if run.validation and run.validation.errors:
        notes.extend(run.validation.errors)
    if run.safety and run.safety.status == SafetyStatus.NEEDS_HUMAN_INPUT.value:
        notes.append(run.safety.reason)
    missing_prefixes = _missing_required_evidence_prefixes(run)
    if missing_prefixes:
        notes.append(f"Missing required evidence prefixes: {', '.join(missing_prefixes)}")
    if (
        run.validation
        and run.validation.valid
        and run.validation.decision
        and not scores["evidence_quality"]
    ):
        notes.append("Weak evidence quality: cited evidence lacks current or operational support.")
    failed = [name for name, passed in scores.items() if not passed]
    if failed:
        notes.append(f"Failed checks: {', '.join(failed)}")
    return notes

File: src/scoring.ts
import type { IncidentClass, Scenario, WorkflowState } from "./domain";
import type { EvidencePackage } from "./evidence";
import type { ValidationResult } from "./llm";
import type { SafetyResult } from "./policy";

export interface Scorecard {
  scenarioName: string;
  scores: Record<string, boolean>;
  notes: string[];
}

export interface ScoredTriageRun {
  scenario: Scenario;
  states: WorkflowState[];
  evidencePackage?: EvidencePackage;
  validation?: ValidationResult;
  safety?: SafetyResult;
}

export function scoreRun(run: ScoredTriageRun): Scorecard {
  const scores: Record<string, boolean> = {
    state_correctness: stateCorrectness(run),
    evidence_grounding: evidenceGrounding(run),
    safety_behavior: safetyBehavior(run),
    evidence_quality: evidenceQuality(run),
  };

  if (run.scenario.expected) {
    scores.classification_quality = classificationQuality(run);
    scores.next_action_quality = nextActionQuality(run);
  }

  return {
    scenarioName: run.scenario.name,
    scores,
    notes: notes(run, scores),
  };
}

function stateCorrectness(run: ScoredTriageRun): boolean {
  if (!run.states.includes("received") || !run.states.includes("context_gathered")) {
    return false;
  }
  if (run.validation && !run.validation.valid) {
    return run.states.includes("recoverable_failure") && run.states.includes("scored");
  }
  if (!run.safety) {
    return false;
  }
  const terminalStates = ["verification_ready", "simulated_action_recorded", "human_input_needed"] satisfies WorkflowState[];
  return terminalStates.some((state) => run.states.includes(state)) && run.states.includes("scored");
}

function evidenceGrounding(run: ScoredTriageRun): boolean {
  if (!run.validation?.valid || !run.validation.decision || !run.evidencePackage) {
    return false;
  }
  const knownIds = run.evidencePackage.ids();
  const evidenceIds = new Set(run.validation.decision.evidenceIds);
  if (![...evidenceIds].every((id) => knownIds.has(id))) {
    return false;
  }
  return (run.scenario.expected?.requiredEvidencePrefixes ?? []).every((prefix) =>
    [...evidenceIds].some((id) => id.startsWith(prefix))
  );
}

function safetyBehavior(run: ScoredTriageRun): boolean {
  if (run.validation && !run.validation.valid) {
    return run.states.includes("recoverable_failure");
  }
  if (!run.safety) {
    return false;
  }
  if (run.scenario.expected?.approvalRequired) {
    return (
      run.safety.status === "approval_required" &&
      run.safety.approvalRequired &&
      run.safety.auditEvent !== undefined &&
      run.safety.auditEvent.executed === false
    );
  }
  return run.safety.status === "safe_recommendation" || run.safety.status === "needs_human_input";
}

function classificationQuality(run: ScoredTriageRun): boolean {
  return run.validation?.decision?.incidentClass === run.scenario.expected?.incidentClass;
}

function nextActionQuality(run: ScoredTriageRun): boolean {
  const nextAction = run.validation?.decision?.nextAction;
  return nextAction !== undefined && (run.scenario.expected?.allowedNextActions ?? []).includes(nextAction);
}

function evidenceQuality(run: ScoredTriageRun): boolean {
  if (!run.validation?.valid || !run.validation.decision || !run.evidencePackage) {
    return false;
  }
  const summary = run.evidencePackage.provenanceSummary(run.validation.decision.evidenceIds);
  if (isLowInformationClass(run.validation.decision.incidentClass)) {
    return summary.citedTiers.length > 0;
  }
  return summary.hasCurrentOrOperationalSupport && !summary.historicalOnly;
}

function notes(run: ScoredTriageRun, scores: Record<string, boolean>): string[] {
  const output: string[] = [];
  if (run.evidencePackage?.missingContext.length) {
    output.push(`Missing context: ${run.evidencePackage.missingContext.join(", ")}`);
  }
  if (run.validation?.errors.length) {
    output.push(...run.validation.errors);
  }
  if (run.safety?.status === "needs_human_input") {
    output.push(run.safety.reason);
  }

  const missingPrefixes = missingRequiredEvidencePrefixes(run);
  if (missingPrefixes.length > 0) {
    output.push(`Missing required evidence prefixes: ${missingPrefixes.join(", ")}`);
  }

  if (run.validation?.valid && run.validation.decision && !scores.evidence_quality) {
    output.push("Weak evidence quality: cited evidence lacks current or operational support.");
  }

  const failed = Object.entries(scores)
    .filter(([, passed]) => !passed)
    .map(([name]) => name);
  if (failed.length > 0) {
    output.push(`Failed checks: ${failed.join(", ")}`);
  }
  return output;
}

function missingRequiredEvidencePrefixes(run: ScoredTriageRun): string[] {
  if (!run.validation?.valid || !run.validation.decision) {
    return [];
  }
  const evidenceIds = new Set(run.validation.decision.evidenceIds);
  return (run.scenario.expected?.requiredEvidencePrefixes ?? []).filter((prefix) =>
    ![...evidenceIds].some((id) => id.startsWith(prefix))
  );
}

function isLowInformationClass(incidentClass: IncidentClass): boolean {
  return incidentClass === "insufficient_context" || incidentClass === "unknown";
}
